### ISSUE TITLE
3.0 regression: gpu process is started despite app.disableHardwareAcceleration() (and offscreen mode)

### ISSUE THREAD

**Issue by bughit** [electron/electron#14273]
* Output of `node_modules/.bin/electron --version`: v3.0.0-beta.7 (x86)
* Operating System (Platform and Version): win7 sp1 x64
* Output of `node_modules/.bin/electron --version` on last known working Electron version (if applicable): v2.0.8 (x86)

**Expected Behavior**
gpu process should not be started, as in 2.0.8
![image](https://user-images.githubusercontent.com/895588/44497970-344c3380-a64a-11e8-92ec-841fd9308ffd.png)

**Actual behavior**
gpu process is started and it exits with error
![image](https://user-images.githubusercontent.com/895588/44498026-6fe6fd80-a64a-11e8-9fdc-e2539a90ec2f.png)
```
[9276:0822/202731.860:ERROR:viz_main_impl.cc(195)] Exiting GPU process due to errors during initialization
[9684:0822/202731.869:ERROR:browser_gpu_channel_host_factory.cc(119)] Failed to launch GPU process.
[9684:0822/202731.869:ERROR:gpu_process_transport_factory.cc(1017)] Lost UI shared context.
```

**To Reproduce**

##### run
`electron.exe --enable-logging offscreen.js`

##### offsreen.js
```js
const { app, BrowserWindow } = require('electron')

const OFFSCREEN = true

if (OFFSCREEN)
  app.disableHardwareAcceleration()

let win

app.once('ready', () => {
  win = new BrowserWindow({
    width: 340,
    height: 240,
    webPreferences: {
      offscreen: OFFSCREEN,
    }
  })

  if (win.webContents.isOffscreen())
    win.webContents.setFrameRate(1)
  
  console.log('%O', {offscreen: win.webContents.isOffscreen(), frameRate: win.webContents.getFrameRate()})
  
  win.webContents.on('paint', (event, dirty, image) => {
    console.log('paint')
  })
  
  win.loadURL('about:blank')
    
  
})
```

**Comment by vladimiry** [comment]
I'm referencing https://github.com/electron/electron/issues/17180 here and closing it then in order to keep a single place for tracking the progress.

@deepak1556 can you add `4-0-x` and `5-0-x` labels to this issue?

**Comment by bughit** [comment]
https://bugs.chromium.org/p/chromium/issues/detail?id=916338

**Comment by nornagon** [comment]
the tl;dr from that Chrome issue is:
- The gpu process isn't just for hardware acceleration. When hw acceleration is disabled, SwiftShader and viz still run in the gpu process. "gpu" is a bit of a misnomer.
- There are a couple of short-lived processes launched during startup that are redundant with --disable-gpu, fixed by  https://crrev.com/960c955a6. That revision does not affect the long-lived "gpu" process.
- If you really, really want to drop processes, you can try `--in-process-gpu`. If you also want to disable SwiftShader (and lose WebGL support), `--disable-software-rasterizer` is there for you.

`--in-process-gpu` is a substantial deviation from the way Chromium generally expects to operate and I expect it to be somewhat less stable than running in the default mode with a separate gpu process (for instance, we would be unable to recover from a crash in the compositor, as that would crash the main process when running with `--in-process-gpu`).

We should make sure we have a JS hook that runs early enough that adding `--in-process-gpu` to the command-line switches is effective, but `app.disableHardwareAcceleration()` should not add that switch by default.

**Comment by vladimiry** [comment]
Just tried `app.commandLine.appendSwitch("in-process-gpu")` and the app fails with one of the scenarios (4.0.6, linux x64):
- Instantly with ```xcb_in.c:746: xcb_request_check: Assertion `!reply' failed.``` message printed to the console.
- After like a second or two with such message printed to the console:
```
[xcb] Unknown sequence number while processing queue
[xcb] Most likely this is a multi-threaded client and XInitThreads has not been called
[xcb] Aborting, sorry about that.
```

**Comment by nornagon** [comment]
Does Chromium work on your machine with the same flag?

**Comment by vladimiry** [comment]
`chromium --in-process-gpu` runs well and I confirm there is no GPU process created.

**Comment by nornagon** [comment]
What happens if you pass `--in-process-gpu` directly on the command-line to electron rather than using `app.commandLine.appendSwitch`?

**Comment by vladimiry** [comment]
> What happens if you pass `--in-process-gpu` directly on the command-line to electron rather than using `app.commandLine.appendSwitch`?

The app starts well and no GPU process created.

Btw adding `app.commandLine.appendSwitch("disable-software-rasterizer")` also disables the GPU process creation and the app doesn't crush.

**Comment by nornagon** [comment]
From my earlier comment:
> We should make sure we have a JS hook that runs early enough that adding `--in-process-gpu` to the command-line switches is effective, but `app.disableHardwareAcceleration()` should not add that switch by default.

This is exactly what I meant. The first JS entry point is too late to properly enable `--in-process-gpu`.

**Comment by MikalDev** [comment]
Working with Electron, Greenworks and Steam in the Windows environment, we need to use --in-process-gpu to enable the Steam overlay. Currently, due to the issue commented above, we need to add the --in-process-gpu command line switch during the launch of the app (e.g. `app.exe --in-process-gpu`). As noted above, if we try to use `app.commandLine.appendSwitch` the app becomes unstable, intermittently crashing.

All this is to say that we also would like to see, so we can add that switch to the code and do not need to rely on adding the command line switch actually on the command line.

"We should make sure we have a JS hook that runs early enough that adding --in-process-gpu to the command-line switches is effective..."

**Comment by vladimiry** [comment]
By the way, just tried electron@5.0.0-beta.6 in action. Adding `disable-software-rasterizer` doesn't help with disabling the GPU process as it worked with electron@4. Adding `in-process-gpu` switch still crushes the app as described above. So the question of how to disable GPU process running electron@5 remains open.

**Comment by dingchaoyan1983** [comment]
--in-process-gpu does not suppered in WINDOWS, the page will not render,  but the IOS platform is ok

**Comment by bughit** [comment]
on win7 with electron 6.0.0-beta.12, either appendSwitch('disable-software-rasterizer') or appendSwitch('in-process-gpu') seem to work

@nornagon Is [special support for in-process-gpu](#issuecomment-468755011) is still needed, perhaps it's only working on windows?

**Comment by nornagon** [comment]
Glad to hear those switches are working for you @bughit. Possibly something in v6 fixed it. I'm going to close this as resolved, will reopen if it recurs.

**Comment by pushkin-** [comment]
@nornagon This came up again ([issue](https://github.com/electron/electron/issues/20702)) fyi. Not sure if it's worth reopening this issue, or just tracking it in the other one.

**Comment by sergeymorkovkin** [comment]
> `--in-process-gpu` does not suppered in WINDOWS, the page will not render, but the IOS platform is ok

Does electon support iOS?

**Comment by RareScrap** [comment]
I was surprised but it still occurs in Electron 35

**Comment by oalfroukh** [comment]
> the tl;dr from that Chrome issue is:
> 
> * The gpu process isn't just for hardware acceleration. When hw acceleration is disabled, SwiftShader and viz still run in the gpu process. "gpu" is a bit of a misnomer.
> * There are a couple of short-lived processes launched during startup that are redundant with --disable-gpu, fixed by  https://crrev.com/960c955a6. That revision does not affect the long-lived "gpu" process.
> * If you really, really want to drop processes, you can try `--in-process-gpu`. If you also want to disable SwiftShader (and lose WebGL support), `--disable-software-rasterizer` is there for you.
> 
> `--in-process-gpu` is a substantial deviation from the way Chromium generally expects to operate and I expect it to be somewhat less stable than running in the default mode with a separate gpu process (for instance, we would be unable to recover from a crash in the compositor, as that would crash the main process when running with `--in-process-gpu`).
> 
> We should make sure we have a JS hook that runs early enough that adding `--in-process-gpu` to the command-line switches is effective, but `app.disableHardwareAcceleration()` should not add that switch by default.

Hello @nornagon 

I was using `app.commandLine.appendSwitch("in-process-gpu");` and `app.commandLine.appendSwitch("disable-software-rasterizer");` on my app to disable the GPU process and it was working fine, while when I upgraded electron  from `v37.x` to `v38.x` and above,I got a black screen, any idea what's going on?

Edited: it's seems that it's a Chromium issue, I just created [issue](https://issues.chromium.org/issues/504284791) on their side
